Cabal/Distribution/Version.hs

```diff
@@ -115,6 +92,95 @@ import Text.PrettyPrint ((<+>))
 import Control.Exception (assert)
 
 -- -----------------------------------------------------------------------------
+-- Versions
+
+-- | A 'Version' represents the version of a software entity.
+--
+-- Instances of 'Eq' and 'Ord' are provided, which gives exact
+-- equality and lexicographic ordering of the version number
+-- components (i.e. 2.1 > 2.0, 1.2.3 > 1.2.2, etc.).
+--
+-- This type is opaque and distinct from the 'Base.Version' type in
+-- "Data.Version" since @Cabal-2.0@. The difference extends to the
+-- 'Binary' instance using a different (and more compact) encoding.
+--
+-- @since 2.0
+data Version = Version [Int]
+             deriving (Data,Eq,Ord,Generic,Show,Read,Typeable)
+
+instance Binary Version
+
+instance NFData Version where
+    rnf = rnf . unVersion
+
+instance Text Version where
+  disp ver
+    = Disp.hcat (Disp.punctuate (Disp.char '.')
+                                (map Disp.int $ unVersion ver))
+
+  parse = do
+      branch <- Parse.sepBy1 parseNat (Parse.char '.')
+                -- allow but ignore tags:
+      _tags  <- Parse.many (Parse.char '-' >> Parse.munch1 isAlphaNum)
+      return (mkVersion branch)
+    where
+      parseNat = read `fmap` Parse.munch1 isDigit
+
+-- | Construct 'Version' from list of version number components.
+--
+-- For instance, @mkVersion [3,2,1]@ constructs a 'Version'
+-- representing the version @3.2.1@.
+--
+-- All version components must be non-negative. @mkVersion []@
+-- represents the special /null/ version; see also 'mkNullVersion' and
+-- 'nullVersion'.
+--
+-- @since 2.0
+mkVersion :: [Int] -> Version
+-- TODO: add validity check; disallow 'mkVersion []' (we have
+-- 'mkNullVersion' for that)
+mkVersion = Version
+
+-- | Variant of 'Version' which converts a "Data.Version" 'Version'
+-- into Cabal's 'Version' type.
+--
+-- @since 2.0
+mkVersion' :: Base.Version -> Version
+mkVersion' = mkVersion . Base.versionBranch
+
+-- | Unpack 'Version' into list of version number components.
+--
+-- This is the inverse to 'mkVersion', so the following holds:
+--
+-- > (unVersion . mkVersion) vs == vs
+--
+-- @since 2.0
+unVersion :: Version -> [Int]
+unVersion (Version vs) = vs
```

Comment:
I was thinking we should give this a name like `versionDigits`.

Once the internal rep changes, it'll be less clear that `unVersion` should refer to producing the abstract representation rather than the underlying physical representation.

Having read through the rest of the code I also think it'd just read a lot better / more clearly.